### PR TITLE
WIP - gen bitswap network

### PR DESCRIPTION
**Pull request by diasdavid**
- [x] func to generate random bitswap networks
- [x] rebase latest @dignifiedquire changes
- [x] confirm that swarm is ok with dialing no protocol at all

### CODE REVIEW

File: test/network/gen-bitswap-network.node.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+      // setTimeout is used to avoid closing the TCP socket while spdy is
+      // still sending a ton of signalling data
+      const tasks = nodeArr.map((node) => {
+        return (cb) => {
+          node.libp2p.swarm.close(cb)
+        }
+      })
+
+      setTimeout(() => {
+        parallel(tasks, done)
```
Comment (dignifiedquire) [inline test/network/gen-bitswap-network.node.js]: Lets use `async.parallel` instead of yet another module in here

Comment (dignifiedquire) [inline test/network/gen-bitswap-network.node.js]: I've been using `run-parallel` only in places where `async` was not used before.

File: test/network/gen-bitswap-network.node.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+      }, 1000)
```
Comment (dignifiedquire) [inline test/network/gen-bitswap-network.node.js]: Would be good to have a comment as to why this timeout exists

Comment (dignifiedquire) [inline test/network/gen-bitswap-network.node.js]: nvm, saw the comment now

File: test/network/gen-bitswap-network.node.js (hunk 1 above)
Comment (diasdavid) [inline test/network/gen-bitswap-network.node.js]: And I thought I would make you happy to stick to your new trend of `run-parallel` :D

File: test/utils.js [hunk 3]
```diff
@@ -84,3 +91,87 @@ exports.applyNetwork = (bs, n) => {
...
+  function createBitswaps () {
+    netArray.forEach((net) => {
+      net.bitswap = new Bitswap(net.peerInfo, net.libp2p, net.repo, net.peerBook)
+    })
+    establishLinks()
+  }
+
+  // connect all the nodes between each other
+  function establishLinks () {
+    async.eachLimit(netArray, 1, (from, cbI) => {
```
Comment (dignifiedquire) [inline test/utils.js]: you can just use `async.series` if you want to limit to `1`

File: test/utils.js [hunk 4]
```diff
@@ -3,7 +3,14 @@
...
 const _ = require('lodash')
 const PeerId = require('peer-id')
+const PeerInfo = require('peer-info')
+const PeerBook = require('peer-book')
+const multiaddr = require('multiaddr')
 const Bitswap = require('../src')
+const libp2p = require('libp2p-ipfs')
+const os = require('os')
+const Repo = require('ipfs-repo')
+const fsBS = require('fs-blob-store')
```
Comment (dignifiedquire) [inline test/utils.js]: Maybe it's more efficient if we use the memory blob store for these tests, what do you think?

File: test/utils.js (hunk 3 above)
Comment (diasdavid) [inline test/utils.js]: good point, changing

**Comment by diasdavid** [comment]
@dignifiedquire applied your CR

File: test/network/gen-bitswap-network.node.js (hunk 1 above)
Comment (dignifiedquire) [inline test/network/gen-bitswap-network.node.js]: sorry :D

**Pull request merged**